che-incubator/che-code
# Move part of logic configuring che-code to javascript instead of keeping it in bash

### What does this PR do?

Adds a simple Typescript project that is compiled to javascript and injected into the editor container.

What does it do? Exactly as the corresponding blocks in the entrypoint:

- configures OpenVSIX registry. If OPENVSX_REGISTRY_URL environment variable is set and not empty, applies it as a registry URL. If OPENVSX_REGISTRY_URL is empty, uses CHE_PLUGIN_REGISTRY_URL as a registry URL.

- if WEBVIEW_LOCAL_RESOURCES environment variable is set to `true`, configures webview static resources to be loaded from the local host (instead of getting them from CDN). It makes webview working when the browser is behind a proxy.

- Creates a bundle certificate and initializes NODE_EXTRA_CERTIFICATE environment variable. The bundle certificate contains content of `/tmp/che/secret/ca.crt` and all found certificates in `/public-certs`.

- Generates `/projects/.code-workspace` file containing all the projects taken from devworkspace flattened devfile.

- Prepares all necessary parameters and launches VS Code.


### What issues does this PR fix?

https://github.com/eclipse/che/issues/22280

### How to test this PR?

**Important:** it is not possible to build PR-based images due to https://github.com/eclipse/che/issues/22390

On the dogfooding instance (due to 12Gb for dev container), create a workspace with https://github.com/che-incubator/che-code/tree/che-code-launcher-test

You should have the workspace with three projects opened.

![Screenshot from 2023-07-07 13-29-19](https://github.com/che-incubator/che-code/assets/1655894/a572d0ed-cace-47e6-af17-6e0182991bc7)
